src/main/java/net/neoforged/neoforge/fluids/crafting/FluidIngredient.java

```diff
@@ -0,0 +1,292 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.fluids.crafting;
+
+import com.mojang.datafixers.util.Either;
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.DataResult;
+import com.mojang.serialization.MapCodec;
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+import net.minecraft.core.Holder;
+import net.minecraft.core.NonNullList;
+import net.minecraft.network.RegistryFriendlyByteBuf;
+import net.minecraft.network.codec.ByteBufCodecs;
+import net.minecraft.network.codec.StreamCodec;
+import net.minecraft.tags.TagKey;
+import net.minecraft.world.item.crafting.Ingredient;
+import net.minecraft.world.level.material.Fluid;
+import net.neoforged.neoforge.common.crafting.ICustomIngredient;
+import net.neoforged.neoforge.common.util.NeoForgeExtraCodecs;
+import net.neoforged.neoforge.fluids.FluidStack;
+import net.neoforged.neoforge.registries.NeoForgeRegistries;
+
+public abstract class FluidIngredient implements Predicate<FluidStack> {
+    /**
+     * This is a codec that is used to represent basic "single fluid" or "tag"
+     * fluid ingredients directly, similar to {@link Ingredient.Value#CODEC},
+     * except not using value subclasses and instead directly providing
+     * the corresponding {@link FluidIngredient}.
+     */
+    private static final MapCodec<FluidIngredient> SINGLE_OR_TAG_CODEC = singleOrTagCodec();
+
+    /**
+     * This is a codec that represents a single {@code FluidIngredient} in map form;
+     * either dispatched by type or falling back to {@link #SINGLE_OR_TAG_CODEC}
+     * if no type is specified.
+     *
+     * @see Ingredient#MAP_CODEC_NONEMPTY
+     */
+    public static final MapCodec<FluidIngredient> MAP_CODEC = makeMapCodec();
```

Comment:
why not call it `MAP_CODEC_NONEMPTY`?

Reply:
Seemed a little verbose (especially since we don't have an empty MapCodec right now), but can do

Reply:
It's just to make it obvious that this is an alternative to `CODEC_NON_EMPTY` rather than `CODEC`.